package/MDAnalysis/lib/util.py

```diff
@@ -1513,23 +1522,54 @@ class A(object):
            @property
            @cached('keyname')
            def size(self):
-               # This code gets ran only if the lookup of keyname fails
-               # After this code has been ran once, the result is stored in
+               # This code gets run only if the lookup of keyname fails
+               # After this code has been run once, the result is stored in
                # _cache with the key: 'keyname'
-               size = 10.0
+               return 10.0
+
+           @property
+           @cached('keyname', universe_validation=True)
+           def othersize(self):
+               # This code gets run only if the lookup
+               # id(self) is not in the validation set under
+               # self.universe._cache['_valid']['keyname']
+               # After this code has been run once, id(self) is added to that
+               # set. The validation set can be centrally invalidated at the
+               # universe level (say, if a topology change invalidates specific
+               # caches).
+               return 20.0
 
 
     .. versionadded:: 0.9.0
 
+    .. versionchanged::2.0.0
+        Added the `universe_validation` keyword.
     """
 
     def cached_lookup(func):
         @wraps(func)
         def wrapper(self, *args, **kwargs):
             try:
+                if universe_validation:  # Universe-level cache validation
+                    u_cache = self.universe._cache.setdefault('_valid', dict())
+                    # A WeakSet is used so that keys from out-of-scope/deleted
+                    # objects don't clutter it.
+                    valid_caches = u_cache.setdefault(key, weakref.WeakSet())
+                    try:
+                        if self._cache_key not in valid_caches:
```

Comment:
As a comment, one day it might be worthwhile to put this `self._cache_key` setting behaviour into `Group.__hash__` so that every AtomGroup natively hashes quickly/cleanly/weakly.

Reply:
Great idea; that seems super clean! I can implement it right away if you're ok with having both functionalities added in the same PR.

Reply:
I went over and checked and we already have a `__hash__` defined, and it looks like it's actually hashing `ix` which isn't going to be fast, but technically necessary for two equivalent AGs to hash identically.  For this lookup I'd rather we quickly hash the AG and maybe have duplicate caches

Reply:
Ah, yea. I hadn't checked `__hash__` but I had initially tried to use the ag itself as the key, and got the same slow performance (I just assumed the entire object was being hashed instead of just `_ix`, which probably boils down to a similar time penalty).